File: content/snippets/go/interview_lcci/05/04_closed_number.go
package main

import "fmt"

func findClosedNumbers(num int) (int, int) {
	bigger, smaller := num, num

	for i := 1; i < 32; i++ {
		if bigger>>(i-1)&1 == 1 && bigger>>i&1 == 0 {
			bigger |= 1 << i
			bigger &^= 1 << (i - 1)
			break
		}
	}

	for i := 1; i < 32; i++ {
		if smaller>>(i-1)&1 == 0 && smaller>>i&1 == 1 {
			smaller &^= 1 << i
			smaller |= 1 << (i - 1)
			break
		}
	}

	return bigger, smaller
}

func main() {
	bigger, smaller := findClosedNumbers(2)
	fmt.Println(bigger)
	fmt.Println(smaller)
}

File: content/snippets/python/interview_lcci/05/04_closed_number.py
def find_closed_numbers(num: int) -> tuple[int, int]:
    bigger = num
    smaller = num

    # Next larger: find rightmost '01' and flip to '10'
    for i in range(1, 32):
        if (bigger >> (i - 1)) & 1 == 1 and (bigger >> i) & 1 == 0:
            bigger |= (1 << i)
            bigger &= ~(1 << (i - 1))
            break

    # Next smaller: find rightmost '10' and flip to '01'
    for i in range(1, 32):
        if (smaller >> (i - 1)) & 1 == 0 and (smaller >> i) & 1 == 1:
            smaller &= ~(1 << i)
            smaller |= (1 << (i - 1))
            break

    return bigger, smaller


def main() -> None:
    bigger, smaller = find_closed_numbers(2)
    print(bigger)
    print(smaller)


if __name__ == "__main__":
    main()
